compiler/rustc_lexer/src/lib.rs

```diff
@@ -151,24 +151,27 @@ pub enum LiteralKind {
     /// "b"abc"", "b"abc"
     ByteStr { terminated: bool },
     /// "r"abc"", "r#"abc"#", "r####"ab"###"c"####", "r#"a"
-    RawStr { n_hashes: u16, err: Option<RawStrError> },
+    RawStr { n_hashes: u16, valid: bool },
     /// "br"abc"", "br#"abc"#", "br####"ab"###"c"####", "br#"a"
-    RawByteStr { n_hashes: u16, err: Option<RawStrError> },
+    RawByteStr { n_hashes: u16, valid: bool },
```

Comment:
Could `n_hashes` be `NonZeroU16` to reduce size such that `bool` could be a niche?

Reply:
No, because `r"a"` is a valid number of hashes.

Reply:
Why not make it `u8`? Would people go crazy and put 255 `#`?

Reply:
It was decided agains making num hashes `u8` in #50296. And tricks like adding one makes the code harder to maintain.
I looked into making `Token` even smaller, but that would need many more adjustments. I guess 8 bytes instead of 16 would be possible, but very complicated.

Reply:
Nice findings, but @nikomatsakis said it was ok https://github.com/rust-lang/rust/pull/50296#issuecomment-385464881.

The reason why `u16` was used is because it was thought that `u8` brings no benefits over `u16`. But here it does bring a benefit, that is reducing the size of token to **8** bytes (without this would be 10 with the other changes I mentioned) which could probably improve compilation time a bit. Maybe CC @rust-lang/lang

Reply:
> Could n_hashes be NonZeroU16 to reduce size such that bool could be a niche?

`(NonZeroU16, bool)` can't space-optimize anyway.

Reply:
Yeah, `NonZero` variant can't space optimize here but `u8` could.